### PR TITLE
feat(model/metrics.py): support calculating accuracy and perplexity m…

### PR DESCRIPTION
**Pull request by huangting4201**
## Motivation

Support calculating model's accuracy and perplexity metrics.

## Modification

Some important updates:
`internlm/model/metrics.py`: new add file `metrics.py` and define `Class AccPerPlex` for calculating model's accuracy and perplexity metrics.
`internlm/core/no_pipeline_scheduler.py`: modify func `forward_backward_step()` for supporting `post_fn` parameter for executing callback function.

The demo example log is shown as follows:
```python
2023-07-25 11:41:00,764 INFO train.py:321 in record_current_batch_training_metrics -- tflops=146.54384242312844 step=15 loss=7.672324180603027 tgs (tokens/gpu/second)=3299.69 lr=1.859423525312737e-05 loss_scale=65536.0 grad_norm=2.3470910228149786 micro_num=4 num_consumed_tokens=1048576 inf_nan_skip_batches=0 num_samples_in_batch=15 largest_length=2048 largest_batch=9 smallest_batch=2 adam_beta2=0.95 fwd_bwd_time=2.02 acc=0.0338 perplexity=2515.8191 acc/en=0.0312 acc/cn=0.0407 acc/code=0.0389 acc/ja=0.0 acc/ar=0.0 acc/kaoshi=0.0 tokens/en=45029 tokens/cn=8185 tokens/code=12287 tokens/ja=0 tokens/ar=0 tokens/kaoshi=0 loss_from_metric=7.8281 loss/en=7.5818 loss/cn=9.237 loss/code=7.7921 loss/ja=nan loss/ar=nan loss/kaoshi=nan 
2023-07-25 11:41:03,361 INFO train.py:321 in record_current_batch_training_metrics -- tflops=142.18926446809743 step=16 loss=6.854360103607178 tgs (tokens/gpu/second)=3201.64 lr=1.490470641152345e-05 loss_scale=65536.0 grad_norm=3.030884081590723 micro_num=4 num_consumed_tokens=1114112 inf_nan_skip_batches=0 num_samples_in_batch=10 largest_length=2048 largest_batch=3 smallest_batch=2 adam_beta2=0.95 fwd_bwd_time=2.09 acc=0.044 perplexity=1409.3516 acc/en=0.0261 acc/cn=0.0311 acc/code=0.0655 acc/ja=0.0 acc/ar=0.0 acc/kaoshi=0.0 tokens/en=28669 tokens/cn=8192 tokens/code=28667 tokens/ja=0 tokens/ar=0 tokens/kaoshi=0 loss_from_metric=7.2535 loss/en=7.0489 loss/cn=8.1349 loss/code=7.2064 loss/ja=nan loss/ar=nan loss/kaoshi=nan
```

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.

### CODE REVIEW

**Comment by huangting4201** [comment]
需要修复冲突，并等pp合并后适配pp

File: train.py [hunk 1]
```diff
@@ -446,10 +460,15 @@ def main(args):
 
         # zero the grads of parameters
         trainer.zero_grad()
+        type_ids = batch[0].pop("type_ids", None)
```
Comment (sunpengsdu) [inline train.py:463]: load_batch的时候，就已经pop了type_ids，所以这里没有

File: train.py [hunk 2]
```diff
@@ -307,6 +307,10 @@ def record_current_batch_training_metrics(
         infos["smallest_batch"] = min_samples_in_batch
         infos["adam_beta2"] = beta2_scheduler.get_beta2()
 
+        res = metric.get_metric()
```
Comment (SolenoidWGT) [inline train.py:310]: metric.get_metric() 里面有dp范围的all-reduce，不要放在 if gpc.is_rank_for_log(): 里面，否则会hang住

Comment (huangting4201) [reply train.py]: 明白了，感谢腾腾！

File: train.py (hunk 1 above)
Comment (huangting4201) [reply train.py:473]: 是的，我把load_batch那里pop type_ids删掉了

**Comment by huangting4201** [comment]
已解决冲突合并develop分支，并完成测试，可review合并

**Review by sunpengsdu — APPROVED** [review]

**Pull request merged**